### PR TITLE
feat: oracle rewards funded by a portion of borrow interest

### PR DESCRIPTION
**Pull request by toteki**
## Description

Oracle rewards funded by a portion of borrow interest. This PR creates the parameter `OracleRewardFactor` and causes rewards to be sent to the `x/oracle` module account periodically.

Note that this occurs every interest epoch, which is likely different than voting period. To address this, `x/oracle` could "smooth out" the differences by only distributing a fraction of the reward pool at a time.

ref: #335 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by adamewozniak** [comment]
re the interest epoch != voting period, the oracle already only rewards (`VotePeriod`/`RewardDistributionWindow`)*`RewardPool` per voting period. Relevant code : 

```go
	distributionRatio := sdk.NewDec(votePeriod).QuoInt64(rewardDistributionWindow)

	var periodRewards sdk.DecCoins
	for _, denom := range rewardDenoms {
		rewardPool := k.GetRewardPool(ctx, denom)

		// return if there's no rewards to give out
		if rewardPool.IsZero() {
			continue
		}

		periodRewards = periodRewards.Add(sdk.NewDecCoinFromDec(
			denom,
			sdk.NewDecFromInt(rewardPool.Amount).Mul(distributionRatio),
		))
	}
```

**Review by alexanderbez — APPROVED** [review]
LGTM! A few notes though:

1. I think we should probably make a comment somewhere in x/leverage spec docs, probably `concepts` doc, about this new functionality. It can be concise.
2. We should also update or tweak slightly the `x/oracle` spec docs to also state this. In addition, we should state when rewards would actually accrue. We can specify a formula or something. E.g. `max(VotingPeriod % IntererstEpoch, ...)`

**Comment by toteki** [comment]
Extra bits in the Oracle and Leverage spec added

**Comment by alexanderbez** [comment]
@adamewozniak can you review the docs changes and suggest anything that you think is relevant or missing please?

**Review by adamewozniak — APPROVED** [review]

File: x/oracle/spec/01_concepts.md [hunk 1]
```diff
@@ -37,6 +37,13 @@ Validators must first pre-commit to a set of exchange rates, then in the subsequ
 
 Let `M` be the weighted median, `𝜎` be the standard deviation of the votes in the ballot, and `R` be the RewardBand parameter. The band around the median is set to be `𝜀 = max(𝜎, R/2)`. All valid (i.e. bonded and non-jailed) validators that submitted an exchange rate vote in the interval `[M - 𝜀, M + 𝜀]` should be included in the set of winners, weighted by their relative vote power.
 
+## Reward Pool
+
+The Oracle module's reward pool is composed of any tokens present in its module account. This pool is funded by the `x/leverage` module as portion of interest accrued on borrowed tokens.
+
+From the Oracle module's perspective, tokens of varied denominations from the `AcceptList` simply appear in the module account at a regular interval.
+The interval is the `x/Leverage` module's `InterestEpoch` parameter, e.g. every 100 blocks, which is generally not equal to the Oracle's `VotePeriod` or other parameters.
+
```
Comment (adamewozniak) [inline x/oracle/spec/01_concepts.md:46]:
```suggestion

The reward pool is not distributed all at once, but instead over a period of time, determined by the param `RewardDistributionWindow`, currently set to `5256000`.
```
```

**Pull request merged**